File: leetcode/20190407/1.py
class Solution:
    def removeOuterParentheses(self, S: str) -> str:
        indices_head, indices_tail = list(), list()
        offset = 0
        while True:
            # S[idx_h:idx_t]
            idx_h, idx_t = self.findPair(S[offset:])
            indices_head.append(offset + idx_h)
            indices_tail.append(offset + idx_t)
            offset = idx_t + offset
            if offset == len(S):
                break
        
        ans = ''
        for idx_h, idx_t in zip(indices_head, indices_tail):
            # to remove ( and )
            ans += S[idx_h + 1: idx_t - 1]
        return ans


    def findPair(self, S: str) -> tuple:
        num = 0
        for i in range(len(S)):

            if S[i] == '(':
                num += 1
                if num == 1:
                    head_idx = i

            elif S[i] == ')':
                num -= 1
                if num == 0:
                    return (head_idx, i + 1)

        raise RuntimeError


# s = '(()())(())'
# s = '(()())(())(()(()))'
# s = '()()'
# print(Solution().removeOuterParentheses(s))

File: leetcode/20190630/1.cpp
class Solution {
public:
    vector<int> distributeCandies(int candies, int num_people) {
        vector<int> ans(num_people, 0);
        int num = 1;
        while (candies != 0) {
            int index = (num - 1) % num_people;
            if (candies <= num) {
                num = candies;
            }
            ans[index] += num;
            candies -= num;
            num++;
        }
        return ans;
    }
};